improver/memprofile.py

```diff
@@ -0,0 +1,125 @@
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module containing maximum memory profiling utilities."""
+
+import tracemalloc as tm
+from queue import Queue, Empty
+from threading import Thread
+from resource import getrusage, RUSAGE_SELF
+from datetime import datetime
+
+
+def memory_profile_start(outfile_prefix, timeout_time=0.1):
+    """Starts the memory tracking profiler.
+
+    Args:
+        outfile_prefix (str):
+            Prefix for the generated output. 2 files will
+            be generated: *_SNAPSHOT and *_MAX_TRACKER.
+        timeout_time (float):
+            Time the queue will look for an input.
+
+    Returns:
+        Active Thread tracking the memory.
+        Active Queue for communication to the thread.
+    """
+    queue = Queue()
+    thread = Thread(target=memory_monitor, args=(queue, timeout_time,
+                                                 outfile_prefix))
+    thread.start()
+    return thread, queue
+
+
+def memory_profile_end(queue, thread):
+    """Ends the memory tracking profiler.
+
+    Args:
+        queue (queue.Queue):
+            Active queue instance to communicate with active thread.
+        thread (thread.Thread):
+            Active thread instance running memory tracking.
+    """
+    queue.put('Stop')
+    thread.join()
+
+
+def memory_monitor(queue, timeout_time, outfile_prefix):
+    """Function to track memory usage, should be run in a separate
+    thread to the main program.
+
+    Args:
+        queue (queue.Queue):
+            Active queue instance to communicate with the thread.
+        timeout_time (float):
+            Time the queue will look for in input.
+        outfile_prefix (str):
+            Prefix for the generated output. 2 files will
+            be generated: *_SNAPSHOT and *_MAX_TRACKER.
+    """
+    tm.start()
```

Comment:
Does the tracemalloc snapshot itself count towards increasing RSS, and always cause a snapshot to be taken in the subsequent checking cycle? Second and subsequent cycles might not trigger that behaviour because the memory space taken for the snapshot (once one exists) is presumably similar.

If my understanding is correct, this function is sampling maxrss at 0.1 second frequency. It's possible (eg. due to spiky memory usage and rapid turnover allocation/deallocation which often happens for each numpy array operation) that the snapshot doesn't actually align with when maxrss occurred - the snapshot will be somewhere between 0.0 and 0.1 seconds later.
That's still probably a reasonably close estimate of maximum memory usage, but it maybe worth mentioning this in a comment so people who use the memory profiles are aware of the situation.

Reply:
Tracemalloc will add to RSS, I'll test whether this means it increases every step, or whether it uses a fixed amount of memory, and comment with the results below.

Your understanding is correct, I'll add a comment explaining it, thanks.

Reply:
Okay, I've tested it by setting a counter, that updates every step and prints out only when max_rss is bigger than old_max - there are gaps in the printed count, so I think it's not just updating every step.